File: yw/dom/Element.lua
--[[
    Copyright (c) 2024, Oh Inseo (YJK) <dhdlstjtr@gmail.com>
    SPDX-License-Identifier: BSD-3-Clause
    This software may contain third-party material. For more info, see README.
]]
local Node                 = require "yw.dom.Node"
local namespaces           = require "yw.common.namespaces"
local HTMLElement          = require "yw.html.HTMLElement"
local HTMLAnchorElement    = require "yw.html.HTMLAnchorElement"
local HTMLBodyElement      = require "yw.html.HTMLBodyElement"
local HTMLBRElement        = require "yw.html.HTMLBRElement"
local HTMLButtonElement    = require "yw.html.HTMLButtonElement"
local HTMLDetailsElement   = require "yw.html.HTMLDetailsElement"
local HTMLDialogElement    = require "yw.html.HTMLDialogElement"
local HTMLDivElement       = require "yw.html.HTMLDivElement"
local HTMLDListElement     = require "yw.html.HTMLDListElement"
local HTMLFieldSetElement  = require "yw.html.HTMLFieldSetElement"
local HTMLFontElement      = require "yw.html.HTMLFontElement"
local HTMLFormElement      = require "yw.html.HTMLFormElement"
local HTMLHeadElement      = require "yw.html.HTMLHeadElement"
local HTMLHeadingElement   = require "yw.html.HTMLHeadingElement"
local HTMLHRElement        = require "yw.html.HTMLHRElement"
local HTMLHtmlElement      = require "yw.html.HTMLHtmlElement"
local HTMLImageElement     = require "yw.html.HTMLImageElement"
local HTMLLIElement        = require "yw.html.HTMLLIElement"
local HTMLLinkElement      = require "yw.html.HTMLLinkElement"
local HTMLMenuElement      = require "yw.html.HTMLMenuElement"
local HTMLMetaElement      = require "yw.html.HTMLMetaElement"
local HTMLObjectElement    = require "yw.html.HTMLObjectElement"
local HTMLOListElement     = require "yw.html.HTMLOListElement"
local HTMLParagraphElement = require "yw.html.HTMLParagraphElement"
local HTMLPreElement       = require "yw.html.HTMLPreElement"
local HTMLQuoteElement     = require "yw.html.HTMLQuoteElement"
local HTMLSpanElement      = require "yw.html.HTMLSpanElement"
local HTMLTableElement     = require "yw.html.HTMLTableElement"
local HTMLTitleElement     = require "yw.html.HTMLTitleElement"
local HTMLUListElement     = require "yw.html.HTMLUListElement"
local object               = require "yw.common.object"
local strings              = require "yw.common.strings"


---https://dom.spec.whatwg.org/#concept-element-custom-element-state
---@alias DOM_Element_CustomElementState "undefined"|"failed"|"uncustomized"|"precustomized"|"custom"

---https://dom.spec.whatwg.org/#concept-element
---@class DOM_Element: DOM_Node
---@field namespace               string?                        https://dom.spec.whatwg.org/#concept-element-namespace
---@field namespacePrefix         string?                        https://dom.spec.whatwg.org/#concept-element-namespace-prefix
---@field localName               string                         https://dom.spec.whatwg.org/#concept-element-local-name
---@field customElementState      DOM_Element_CustomElementState https://dom.spec.whatwg.org/#concept-element-custom-element-state
---@field customElementDefinition HTML_CustomElementDefinition?  https://dom.spec.whatwg.org/#concept-element-custom-element-definition
---@field is                      string?                        https://dom.spec.whatwg.org/#concept-element-is-value
---@field attributes              DOM_Attr                       https://dom.spec.whatwg.org/#concept-element-attribute
---@field shadowRoot              DOM_Node?                      https://dom.spec.whatwg.org/#concept-element-shadow-root
local Element = object.create(Node)

---@param attributes DOM_Attr[]
---@param namespace string?
---@param namespacePrefix string?
---@param localName string
---@param customElementState DOM_Element_CustomElementState
---@param customElementDefinition HTML_CustomElementDefinition?
---@param is string?
---@param document DOM_Document
---@return DOM_Element
function Element:new(attributes, namespace, namespacePrefix, localName, customElementState, customElementDefinition, is,
                     document)
    local o = Node.new(self, document) --[[@as DOM_Element]]
    o.isElement = true
    o.shadowRoot = nil
    o.attributes = attributes
    o.namespace = namespace
    o.namespacePrefix = namespacePrefix
    o.localName = localName
    o.customElementState = customElementState
    o.customElementDefinition = customElementDefinition
    o.is = is
    return o
end

local HTML_ELEMENT_INTERFACES = {
    ["a"]          = HTMLAnchorElement,
    ["blockquote"] = HTMLQuoteElement,
    ["body"]       = HTMLBodyElement,
    ["br"]         = HTMLBRElement,
    ["button"]     = HTMLButtonElement,
    ["details"]    = HTMLDetailsElement,
    ["dialog"]     = HTMLDialogElement,
    ["div"]        = HTMLDivElement,
    ["dl"]         = HTMLDListElement,
    ["fieldset"]   = HTMLFieldSetElement,
    ["font"]       = HTMLFontElement,
    ["form"]       = HTMLFormElement,
    ["h1"]         = HTMLHeadingElement,
    ["h2"]         = HTMLHeadingElement,
    ["h3"]         = HTMLHeadingElement,
    ["h4"]         = HTMLHeadingElement,
    ["h5"]         = HTMLHeadingElement,
    ["h6"]         = HTMLHeadingElement,
    ["head"]       = HTMLHeadElement,
    ["hr"]         = HTMLHRElement,
    ["html"]       = HTMLHtmlElement,
    ["img"]        = HTMLImageElement,
    ["li"]         = HTMLLIElement,
    ["link"]       = HTMLLinkElement,
    ["menu"]       = HTMLMenuElement,
    ["meta"]       = HTMLMetaElement,
    ["object"]     = HTMLObjectElement,
    ["ol"]         = HTMLOListElement,
    ["p"]          = HTMLParagraphElement,
    ["pre"]        = HTMLPreElement,
    ["span"]       = HTMLSpanElement,
    ["table"]      = HTMLTableElement,
    ["title"]      = HTMLTitleElement,
    ["ul"]         = HTMLUListElement,
    ----------------------------------------------------------------------------
    ["address"]    = HTMLElement,
    ["article"]    = HTMLElement,
    ["aside"]      = HTMLElement,
    ["b"]          = HTMLElement,
    ["cite"]       = HTMLElement,
    ["code"]       = HTMLElement,
    ["dd"]         = HTMLElement,
    ["dfn"]        = HTMLElement,
    ["dt"]         = HTMLElement,
    ["em"]         = HTMLElement,
    ["figcaption"] = HTMLElement,
    ["figure"]     = HTMLElement,
    ["footer"]     = HTMLElement,
    ["header"]     = HTMLElement,
    ["hgroup"]     = HTMLElement,
    ["i"]          = HTMLElement,
    ["main"]       = HTMLElement,
    ["nav"]        = HTMLElement,
    ["s"]          = HTMLElement,
    ["search"]     = HTMLElement,
    ["section"]    = HTMLElement,
    ["small"]      = HTMLElement,
    ["strong"]     = HTMLElement,
    ["summary"]    = HTMLElement,
    ["u"]          = HTMLElement,
    ["var"]        = HTMLElement,
}

-- https://dom.spec.whatwg.org/#concept-element-interface
---@param localName string
---@param namespace string?
---@return DOM_Element
local function elementInterface(localName, namespace)
    local interfaces = nil
    if namespace == namespaces.HTML_NAMESPACE then
        interfaces = HTML_ELEMENT_INTERFACES
    elseif namespace == namespaces.SVG_NAMESPACE then
        error("TODO")
    elseif namespace == namespaces.MATHML_NAMESPACE then
        error("TODO")
    end
    if interfaces ~= nil then
        for name, interface in pairs(interfaces) do
            if name == localName then
                return interface
            end
        end
    end
    -- The element interface for any name and namespace is Element, unless stated otherwise.
    return Element
end


---https://dom.spec.whatwg.org/#concept-create-element
---@param document DOM_Document
---@param localName string
---@param namespace string?
---@param namespacePrefix string?
---@param is string?
---@param synchronousCustomElements? boolean
---@return DOM_Element
function Element.create(document, localName, namespace, namespacePrefix, is, synchronousCustomElements)
    if synchronousCustomElements == nil then synchronousCustomElements = false end

    --- 1. Let result be null.
    local result = nil

    -- 2. Let definition be the result of looking up a custom element definition given document, namespace, localName, and is.
    local definition = nil
    -- TODO

    -- 3. If definition is non-null, and definition’s name is not equal to its local name (i.e., definition represents a customized built-in element):
    if definition ~= nil then
        error("TODO")

        -- 1. Let interface be the element interface for localName and the HTML namespace.

        -- 2. Set result to a new element that implements interface, with no attributes, namespace set to the HTML namespace, namespace prefix set to prefix, local name set to localName, custom element state set to "undefined", custom element definition set to null, is value set to is, and node document set to document.

        -- 3. If synchronousCustomElements is true, then run this step while catching any exceptions:

        --        1. Upgrade result using definition.

        --    If this step threw an exception exception:

        --        1. Report exception for definition’s constructor’s corresponding JavaScript object’s associated realm’s global object.

        --        2. Set result’s custom element state to "failed".

        -- 4. Otherwise, enqueue a custom element upgrade reaction given result and definition.
    elseif definition ~= nil then --[[
        4. Otherwise, if definition is non-null:
    ]]
        -- 1. If synchronousCustomElements is true, then run these steps while catching any exceptions:

        --        1. Let C be definition’s constructor.

        --        2. Set result to the result of constructing C, with no arguments.

        --        3. Assert: result’s custom element state and custom element definition are initialized.

        --        4. Assert: result’s namespace is the HTML namespace.

        --        5. If result’s attribute list is not empty, then throw a "NotSupportedError" DOMException.

        --        6. If result has children, then throw a "NotSupportedError" DOMException.

        --        7. If result’s parent is not null, then throw a "NotSupportedError" DOMException.

        --        8. If result’s node document is not document, then throw a "NotSupportedError" DOMException.

        --        9. If result’s local name is not equal to localName, then throw a "NotSupportedError" DOMException.

        --        10. Set result’s namespace prefix to prefix.

        --        11. Set result’s is value to null.

        --    If any of these steps threw an exception exception:

        --        1. Report exception for definition’s constructor’s corresponding JavaScript object’s associated realm’s global object.

        --        2. Set result to a new element that implements the HTMLUnknownElement interface, with no attributes, namespace set to the HTML namespace, namespace prefix set to prefix, local name set to localName, custom element state set to "failed", custom element definition set to null, is value set to null, and node document set to document.

        -- 2. Otherwise:

        --        1. Set result to a new element that implements the HTMLElement interface, with no attributes, namespace set to the HTML namespace, namespace prefix set to prefix, local name set to localName, custom element state set to "undefined", custom element definition set to null, is value set to null, and node document set to document.

        --        2. Enqueue a custom element upgrade reaction given result and definition.
    else --[[
        5. Otherwise:
    ]]
        -- 1. Let interface be the element interface for localName and namespace.
        local interface = elementInterface(localName, namespace)

        -- 2. Set result to a new element that implements interface,
        result = interface:new(
            {},              -- with no attributes,
            namespace,       -- namespace set to namespace,
            namespacePrefix, -- namespace prefix set to prefix,
            localName,       -- local name set to localName,
            "uncustomized",  -- custom element state set to "uncustomized",
            nil,             -- custom element definition set to null,
            is,              -- is value set to is,
            document         -- and node document set to document.
        )

        -- 3. If namespace is the HTML namespace,
        if namespace == namespaces.HTML_NAMESPACE and
            -- and either localName is a valid custom element name
            (true --[[ TODO ]] or
                -- or is is non-null,
                is ~= nil
            )
        then
            -- then set result’s custom element state to "undefined".
            result.customElementState = "undefined"
        end
    end
    assert(result ~= nil)
    -- 6. Return result.
    return result
end

---https://dom.spec.whatwg.org/#element-shadow-host
---@return boolean
function Element:isShadowHost()
    -- An element is a shadow host if its shadow root is non-null.
    return self.shadowRoot ~= nil
end

---https://dom.spec.whatwg.org/#concept-element-qualified-name
---@return string
function Element:qualifiedName()
    -- An element’s qualified name is
    if self.namespacePrefix == nil then
        -- its local name if its namespace prefix is null;
        return self.localName
    end
    -- otherwise its namespace prefix, followed by ":", followed by its local name.
    return self.namespacePrefix .. ":" .. self.localName
end

---https://dom.spec.whatwg.org/#element-html-uppercased-qualified-name
---@return string
function Element:htmlUppercasedQualifiedName()
    -- 1. Let qualifiedName be this’s qualified name.
    local qualifiedName = self:qualifiedName()
    -- 2. If this is in the HTML namespace and its node document is an HTML document, then set qualifiedName to qualifiedName in ASCII uppercase.
    if (self.namespace == namespaces.HTML_NAMESPACE) and (self.nodeDocument.type == "html") then
        qualifiedName = strings.asciiLowercase(qualifiedName)
    end
    -- 3. Return qualifiedName.
    return qualifiedName
end

---https://dom.spec.whatwg.org/#concept-element-custom
---@return boolean
function Element:isCustom()
    -- An element whose custom element state is "custom" is said to be custom.
    return self.customElementState == "custom"
end

---https://dom.spec.whatwg.org/#concept-element-attributes-append
---@param attribute DOM_Attr
function Element:appendAttribute(attribute)
    -- 1. Append attribute to element’s attribute list.
    table.insert(self.attributes, attribute)
    -- 2. Set attribute’s element to element.
    attribute.element = self
    -- 3. Handle attribute changes for attribute with element, null, and attribute’s value.
    -- TODO
end

---@param name string
---@param namespace string?
---@param prefix string?
---@return DOM_Attr|nil
function Element:getAttribute(name, namespace, prefix)
    for _, attr in ipairs(self.attributes) do
        if attr.localName == name
            and (namespace == nil or attr.namespace == namespace)
            and (prefix == nil or attr.prefix == prefix)
        then
            return attr
        end
    end
    return nil
end

---@param name string
---@param namespace string?
---@param prefix string?
---@return string|nil
function Element:getAttributeValue(name, namespace, prefix)
    local attr = self:getAttribute(name, namespace, prefix)
    if attr == nil then
        return nil
    end
    return attr.value
end

---@param name string
---@param value string
---@param namespace string?
---@param prefix string?
function Element:setAttributeValue(name, value, namespace, prefix)
    local attr = self:getAttribute(name, namespace, prefix)
    if attr == nil then
        error("No such attribute: " .. tostring(name))
    end
    attr.value = value
end

---@param localName string?
---@return boolean
function Element:isHTMLElement(localName)
    if self.namespace ~= namespaces.HTML_NAMESPACE then
        return false
    end
    if localName ~= nil and self.localName ~= localName then
        return false
    end
    return true
end

---@param localName string?
---@return boolean
function Element:isSVGElement(localName)
    if self.namespace ~= namespaces.SVG_NAMESPACE then
        return false
    end
    if localName ~= nil and self.localName ~= localName then
        return false
    end
    return true
end

---@param localName string?
---@return boolean
function Element:isMathMLElement(localName)
    if self.namespace ~= namespaces.MATHML_NAMESPACE then
        return false
    end
    if localName ~= nil and self.localName ~= localName then
        return false
    end
    return true
end

---@param localNames string[]
---@return boolean
function Element:isOneOfHTMLElements(localNames)
    if self.namespace ~= namespaces.HTML_NAMESPACE then
        return false
    end
    for _, n in ipairs(localNames) do
        if n == self.localName then
            return true
        end
    end
    return false
end

local SPECIAL_CATEGORY_HTML_TAG_NAMES = {
    "address", "applet", "area", "article", "aside", "base", "basefont",
    "bgsound", "blockquote", "body", "br", "button", "caption", "center",
    "col", "colgroup", "dd", "details", "dir", "div", "dl", "dt", "embed",
    "fieldset", "figcaption", "figure", "footer", "form", "frame", "frameset",
    "h1", "h2", "h3", "h4", "h5", "h6", "head", "header", "hgroup", "hr",
    "html", "iframe", "img", "input", "keygen", "li", "link", "listing", "main",
    "marquee", "menu", "meta", "nav", "noembed", "noframes", "noscript",
    "object", "ol", "p", "param", "plaintext", "pre", "script", "search",
    "section", "select", "source", "style", "summary", "table", "tbody", "td",
    "template", "textarea", "tfoot", "th", "thead", "title", "tr", "track",
    "ul", "wbr", "xmp"
}
local SPECIAL_CATEGORY_MATHML_TAG_NAMES = { "mi", "mn", "ms", "mtext", "annotation-xml" }
local SPECIAL_CATEGORY_SVG_TAG_NAMES = { "foreignObject", "desc", "title" }

---https://html.spec.whatwg.org/multipage/parsing.html#special
---@return boolean
function Element:isInSpecialCategory()
    for _, t in ipairs(SPECIAL_CATEGORY_HTML_TAG_NAMES) do
        if self:isHTMLElement(t) then
            return true
        end
    end
    for _, t in ipairs(SPECIAL_CATEGORY_MATHML_TAG_NAMES) do
        if self:isMathMLElement(t) then
            return true
        end
    end
    for _, t in ipairs(SPECIAL_CATEGORY_SVG_TAG_NAMES) do
        if self:isSVGElement(t) then
            return true
        end
    end
    return false
end

---https://dom.spec.whatwg.org/#dom-element-tagname
---@return string
function Element:tagName()
    -- The tagName getter steps are to return this’s HTML-uppercased qualified name.
    return self:htmlUppercasedQualifiedName()
end

return Element

File: dom/element.go
package dom

import (
	"fmt"
	"log"
	"slices"
	"strconv"
	"strings"

	"github.com/inseo-oh/yw/namespaces"
	"github.com/inseo-oh/yw/util"
)

// Element represents a [DOM element].
//
// TODO(ois): Should we move some of Element's HTML related extensions to the html package instead?
//
// [DOM element]: https://dom.spec.whatwg.org/#concept-element
type Element interface {
	Node

	// Namespace returns [namespace] of the element. ok is set to false if it's absent.
	//
	// [namespace]: https://dom.spec.whatwg.org/#concept-element-namespace
	Namespace() (ns namespaces.Namespace, ok bool)

	// Prefix returns [namespace prefix] of the element. ok is set to false if it's absent.
	//
	// TODO(ois): Should we call this NamespacePrefix instead?
	// [namespace prefix]: https://dom.spec.whatwg.org/#concept-element-namespace-prefix.
	Prefix() (pr string, ok bool)

	// LocalName returns [local name] of the element.
	//
	// [local name]: https://dom.spec.whatwg.org/#concept-element-local-name
	LocalName() string

	// CustomElementRegistry returns [custom element registry] of the element.
	//
	// [custom element registry]: https://dom.spec.whatwg.org/#element-custom-element-registry
	CustomElementRegistry() *CustomElementRegistry

	// CustomElementRegistry sets [custom element registry] of the element to reg.
	//
	// [custom element registry]: https://dom.spec.whatwg.org/#element-custom-element-registry
	SetCustomElementRegistry(reg *CustomElementRegistry)

	// CustomElementRegistry returns [is value] of the element.
	//
	// [is value]: https://dom.spec.whatwg.org/#concept-element-is-value
	Is() (string, bool)

	// IsCustom reports whether element is [custom].
	//
	// [custom]: https://dom.spec.whatwg.org/#concept-element-custom
	IsCustom() bool

	// ShadowRoot returns [shadow root] of the element.
	//
	// [shadow root]: https://dom.spec.whatwg.org/#concept-element-shadow-root
	ShadowRoot() ShadowRoot

	// SetShadowRoot sets [shadow root] of the element to sr.
	//
	// [shadow root]: https://dom.spec.whatwg.org/#concept-element-shadow-root
	SetShadowRoot(sr ShadowRoot)

	// IsShadowHost reports whether element is [shadow host].
	//
	// [shadow host]: https://dom.spec.whatwg.org/#element-shadow-host
	IsShadowHost() bool

	// Attrs returns [attributes] of the element.
	//
	// [attributes]: https://dom.spec.whatwg.org/#concept-element-attribute
	Attrs() []Attr

	// Attrs appends new attribute to [attributes] of the element.
	//
	// [attributes]: https://dom.spec.whatwg.org/#concept-element-attribute
	AppendAttr(attrData AttrData)

	// AttrWithNamespace searches attribute from element's [attributes], that
	// matches namePair's Namespace and Name, and returns its value.
	// Attributes without namespace are ignored. ok is set to false if there's
	// no such attribute.
	//
	// [attributes]: https://dom.spec.whatwg.org/#concept-element-attribute
	AttrWithNamespace(namePair NamePair) (value string, ok bool)

	// AttrWithoutNamespace searches attribute from element's [attributes], that
	// matches the name, and returns its value.
	// Attributes with namespace are ignored. ok is set to false if there's no
	// such attribute.
	//
	// [attributes]: https://dom.spec.whatwg.org/#concept-element-attribute
	AttrWithoutNamespace(name string) (value string, ok bool)

	// IsElement reports whether the element's local name is equal to namePair's
	// LocalName and is in name's Namespace.
	IsElement(namePair NamePair) bool

	// IsHtmlElement reports whether the element's local name is equal to given
	// localName and is in HTML namespace.
	IsHtmlElement(localName string) bool

	// IsMathmlElement reports whether the element's local name is equal to given
	// localName and is in MathML namespace.
	IsMathmlElement(localName string) bool

	// IsSvgElement reports whether the element's local name is equal to given
	// localName and is in SVG namespace.
	IsSvgElement(localName string) bool

	// IntrinsicSize returns intrinsic size of the element.
	IntrinsicSize() (width float64, height float64)

	// IsInside reports whether the element is inside another element that matches
	// given namePair's Namespace and Name.
	IsInside(namePair NamePair) bool

	// TagToken returns associated HTML tag token for the element.
	TagToken() TagToken

	//==========================================================================
	// Some extensions for HTML parser
	//==========================================================================

	// Reports whether the element is in HTML parser's [special category].
	//
	// [special category]: https://html.spec.whatwg.org/multipage/parsing.html#special
	IsHtmlSpecialElement() bool

	// Reports whether the element is in HTML parser's [formatting category].
	//
	// [formatting category]: https://html.spec.whatwg.org/multipage/parsing.html#formatting
	IsHtmlFormattingElement() bool

	// Reports whether the element is in HTML parser's [ordinary category].
	//
	// [ordinary category]: https://html.spec.whatwg.org/multipage/parsing.html#ordinary
	IsHtmlOrdinaryElement() bool

	// Reports whether the element is [MathML text integration point].
	//
	// [MathML text integration point]: https://html.spec.whatwg.org/multipage/parsing.html#mathml-text-integration-point
	IsMathmlTextIntegrationPoint() bool

	// Reports whether the element is [HTML integration point].
	//
	// [HTML integration point]: https://html.spec.whatwg.org/multipage/parsing.html#html-integration-point
	IsHtmlIntegrationPoint() bool
}

type elementImpl struct {
	Node

	namespace             *namespaces.Namespace // May be nil
	prefix                *string               // May be nil
	localName             string
	is                    *string                // May be nil
	customElementReigstry *CustomElementRegistry // May be nil
	shadowRoot            ShadowRoot             // may be nil
	attrs                 []Attr
	tagToken              TagToken
	customElementState    CustomElementState
}
type CustomElementState uint8

// https://dom.spec.whatwg.org/#concept-element-custom-element-state
const (
	CustomElementUndefined CustomElementState = iota
	CustomElementFailed
	CustomElementUncustomized
	CustomElementPrecustomized
	CustomElementCustom
)

func NewElement(options ElementCreationCommonOptions) Element {
	return &elementImpl{
		Node:                  NewNode(options.NodeDocument),
		namespace:             options.Namespace,
		prefix:                options.Prefix,
		localName:             options.LocalName,
		attrs:                 []Attr{},
		tagToken:              options.TagToken,
		customElementReigstry: options.CustomElementRegistry,
		customElementState:    options.CustomElementState,
		is:                    options.Is,
	}
}
func (n elementImpl) LocalName() string { return n.localName }
func (n elementImpl) Namespace() (namespaces.Namespace, bool) {
	if n.namespace == nil {
		return "", false
	}
	return *n.namespace, true
}
func (n elementImpl) Prefix() (string, bool) {
	if n.prefix == nil {
		return "", false
	}
	return *n.prefix, true
}
func (n elementImpl) Is() (string, bool) {
	if n.is == nil {
		return "", false
	}
	return *n.is, true
}
func (n elementImpl) CustomElementRegistry() *CustomElementRegistry {
	return n.customElementReigstry
}
func (n *elementImpl) SetCustomElementRegistry(reg *CustomElementRegistry) {
	n.customElementReigstry = reg
}
func (n elementImpl) ShadowRoot() ShadowRoot {
	return n.shadowRoot
}
func (n *elementImpl) SetShadowRoot(sr ShadowRoot) {
	n.shadowRoot = sr
}
func (n elementImpl) IsShadowHost() bool {
	return !util.IsNil(n.shadowRoot)
}
func (n elementImpl) IsDefined() bool {
	// https://dom.spec.whatwg.org/#concept-element-defined
	return n.customElementState == CustomElementUncustomized ||
		n.customElementState == CustomElementCustom
}
func (n elementImpl) IsCustom() bool {
	// https://dom.spec.whatwg.org/#concept-element-custom
	return n.customElementState == CustomElementCustom
}
func (n elementImpl) Attrs() []Attr {
	return n.attrs
}
func (n elementImpl) TagToken() TagToken {
	return n.tagToken
}
func (n elementImpl) IntrinsicSize() (width float64, height float64) {
	if n.Callbacks().IntrinsicSize == nil {
		log.Printf("%v does not implement IntrinsicSize(). Assuming size 0, 0", n)
		return 0, 0
	}
	return n.Callbacks().IntrinsicSize()
}

func (n elementImpl) String() string {
	sb := strings.Builder{}
	sb.WriteString("<")
	if n.namespace != nil {
		sb.WriteString(fmt.Sprintf("%v:", n.namespace))
	}
	sb.WriteString(n.localName)
	for _, attr := range n.attrs {
		sb.WriteString(" ")
		val := strconv.Quote(attr.Value())
		if ns, ok := attr.Namespace(); ok {
			sb.WriteString(fmt.Sprintf("%v:%s=%s", ns, attr.LocalName(), val))
		} else {
			sb.WriteString(fmt.Sprintf("%s=%s", attr.LocalName(), val))
		}
	}
	sb.WriteString(">")
	return sb.String()
}

type NamePair struct {
	Namespace namespaces.Namespace
	LocalName string
}

func (n elementImpl) IsInside(namePair NamePair) bool {
	current := n.Parent()
	for !util.IsNil(current) {
		currElem, ok := current.(Element)
		if !ok {
			break
		} else if currElem.IsElement(namePair) {
			return true
		}
		current = currElem.(Node).Parent()
	}
	return false
}

// Note that attr.element will be overwritten by this function, so don't set it.
func (n *elementImpl) AppendAttr(attrData AttrData) {
	attr := NewAttr(attrData.LocalName, attrData.Value, attrData.Namespace, attrData.NamespacePrefix, n)
	n.attrs = append(n.attrs, attr)
}

func (n elementImpl) IsElement(namePair NamePair) bool {
	return n.namespace != nil && *n.namespace == namePair.Namespace && n.localName == namePair.LocalName
}
func (n elementImpl) IsHtmlElement(localName string) bool {
	return n.IsElement(NamePair{namespaces.Html, localName})
}
func (n elementImpl) IsMathmlElement(localName string) bool {
	return n.IsElement(NamePair{namespaces.Mathml, localName})
}
func (n elementImpl) IsSvgElement(localName string) bool {
	return n.IsElement(NamePair{namespaces.Svg, localName})
}

func (n elementImpl) AttrWithNamespace(namePair NamePair) (string, bool) {
	for _, attr := range n.attrs {
		if ns, hasNs := attr.Namespace(); hasNs && ns == namePair.Namespace && attr.LocalName() == namePair.LocalName {
			return attr.Value(), true
		}
	}
	return "", false
}

// Returns nil if not found. This will only match attributes without namespace specified.
func (n elementImpl) AttrWithoutNamespace(name string) (string, bool) {
	for _, attr := range n.attrs {
		if _, hasNs := attr.Namespace(); !hasNs && attr.LocalName() == name {
			return attr.Value(), true
		}
	}
	return "", false
}

//------------------------------------------------------------------------------
// Some extensions for HTML parser
//------------------------------------------------------------------------------

// https://html.spec.whatwg.org/multipage/parsing.html#special
func (n elementImpl) IsHtmlSpecialElement() bool {
	specialHtmlElems := []string{
		"address", "applet", "area", "article", "aside", "base", "basefont",
		"bgsound", "blockquote", "body", "br", "button", "caption", "center",
		"col", "colgroup", "dd", "details", "dir", "div", "dl", "dt", "embed",
		"fieldset", "figcaption", "figure", "footer", "form", "frame",
		"frameset", "h1", "h2", "h3", "h4", "h5", "h6", "head", "header",
		"hgroup", "hr", "html", "iframe", "img", "input", "keygen", "li",
		"link", "listing", "main", "marquee", "menu", "meta", "nav", "noembed",
		"noframes", "noscript", "object", "ol", "p", "param", "plaintext",
		"pre", "script", "search", "section", "select", "source", "style",
		"summary", "table", "tbody", "td", "template", "textarea", "tfoot",
		"th", "thead", "title", "tr", "track", "ul", "wbr", "xmp",
	}
	specialMathmlElems := []string{
		"mi", "mo", "mn", "ms", "mtext", "annotation-xml",
	}
	specialSvgElems := []string{
		"foreignObject", "desc", "title",
	}
	if slices.ContainsFunc(specialHtmlElems, n.IsHtmlElement) {
		return true
	}
	if slices.ContainsFunc(specialMathmlElems, n.IsMathmlElement) {
		return true
	}
	return slices.ContainsFunc(specialSvgElems, n.IsSvgElement)
}

// https://html.spec.whatwg.org/multipage/parsing.html#formatting
func (n elementImpl) IsHtmlFormattingElement() bool {
	specialHtmlElems := []string{
		"a", "b", "big", "code", "em", "font", "i", "nobr", "s", "small",
		"strike", "strong", "tt", "u",
	}
	return slices.ContainsFunc(specialHtmlElems, n.IsHtmlElement)
}

// https://html.spec.whatwg.org/multipage/parsing.html#ordinary
func (n elementImpl) IsHtmlOrdinaryElement() bool {
	return !n.IsHtmlSpecialElement() && !n.IsHtmlFormattingElement()
}

// https://html.spec.whatwg.org/multipage/parsing.html#mathml-text-integration-point
func (n elementImpl) IsMathmlTextIntegrationPoint() bool {
	return slices.ContainsFunc([]string{"mi", "mo", "mn", "ms", "mtext"}, n.IsMathmlElement)
}

// https://html.spec.whatwg.org/multipage/parsing.html#html-integration-point
func (n elementImpl) IsHtmlIntegrationPoint() bool {
	if n.IsMathmlElement("annotation-xml") {
		if attr, ok := n.tagToken.Attr("encoding"); ok &&
			(util.ToAsciiLowercase(attr) == "text/html" || util.ToAsciiLowercase(attr) == "application/xhtml+xml") {
			return true
		}
	}
	if n.IsSvgElement("foreignObject") ||
		n.IsSvgElement("desc") ||
		n.IsSvgElement("title") {
		return true
	}
	return false
}
